src/util/hash.rs

```diff
@@ -25,48 +25,50 @@ use core::cmp::min;
 use hashes::Hash;
 use consensus::encode::Encodable;
 
-/// Calculates the merkle root of a list of hashes inline
-/// into the allocated slice.
+/// Calculates the merkle root of a list of hashes inline (in place) in `data`.
 ///
 /// In most cases, you'll want to use [bitcoin_merkle_root] instead.
+///
+/// # Panics
+/// Panics if, and only if, `data` is empty. This function should only ever be called with the
+/// transaction data from a block, which should never be empty because blocks always contain at
+/// least one transaction (the coinbase transaction).
 pub fn bitcoin_merkle_root_inline<T>(data: &mut [T]) -> T
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
 {
-    // Base case
-    if data.is_empty() {
-        return Default::default();
-    }
-    if data.len() < 2 {
-        return T::from_inner(data[0].into_inner());
+    if data.len() == 0 {
+        panic!("merkle root of an empty tree is undefined");
     }
-    // Recursion
-    for idx in 0..((data.len() + 1) / 2) {
-        let idx1 = 2 * idx;
-        let idx2 = min(idx1 + 1, data.len() - 1);
-        let mut encoder = T::engine();
-        data[idx1].consensus_encode(&mut encoder).unwrap();
-        data[idx2].consensus_encode(&mut encoder).unwrap();
-        data[idx] = T::from_engine(encoder);
+
+    // If there is only a single hash then that hash is, by definition, the merkle root.
+    if data.len() == 1 {
+        return data[0];
     }
-    let half_len = data.len() / 2 + data.len() % 2;
-    bitcoin_merkle_root_inline(&mut data[0..half_len])
+
+    merkle_root_r(data)
 }
 
 /// Calculates the merkle root of an iterator of hashes.
+///
+/// # Panics
+/// The merkle root of an empty tree of hashes is undefined, panics if `iter` is empty.
 pub fn bitcoin_merkle_root<T, I>(mut iter: I) -> T
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
           I: ExactSizeIterator<Item = T>,
```

Comment:
Unrelated but I think `ExactSIzeIterator` is not actually required. Just allocate `Vec` based on `size_hint` and let the `Vec` resize itself if the hint is incorrect (that's actually what happens today - len is not strictly guaranteed to be correct).

Reply:
But then we cannot call `iter.len()`?

Reply:
I don't think we need to:

```rust
let first = iter.next()?;
let second = match iter.next() {
    Some(second) => second,
    // A single hash is by definition the merkle root.
    None => return Some(first),
};
let (min, max) = iter.size_hint();
let mut hashes = Vec::with_capacity(max.unwrap_or(min) / 2 + 1);
let iter = [first, second].iter().chain(iter);
// Continue using `hashes` and `iter` as before.
```

Reply:
> let iter = [first, second].iter().chain(iter);

I couldn't get this line to build, I used instead:
```
let mut hashes = iter::once(first).chain(iter::once(second)).chain(hashes);
```